From c1710c84c0a563ccbe86dba427d7584c13c7ff1b Mon Sep 17 00:00:00 2001
From: Ido Ben-Hur <idoybh2@gmail.com>
Date: Mon, 18 Dec 2023 17:05:03 +0200
Subject: [PATCH 090/103] Screenrecord: Fix notifications not being dismissed

...after using actions such as "delete" or "share".
Post / Recieve the notification ID with the intent for that.
Also count grouped notifications to decide when to dismiss the group
summary and only close QS after all notifications are dismissed.

Change-Id: I1023367412b2b737b1cf3e9935b1d268059a2a3d
---
 .../screenrecord/RecordingService.java        | 86 +++++++++++++++----
 1 file changed, 68 insertions(+), 18 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java b/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
index ef91a5708ece..7c1d35452736 100644
--- a/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
+++ b/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
@@ -37,6 +37,7 @@ import android.os.RemoteException;
 import android.os.SystemClock;
 import android.os.UserHandle;
 import android.provider.Settings;
+import android.service.notification.StatusBarNotification;
 import android.util.Log;
 import android.view.animation.AlphaAnimation;
 import android.view.animation.Animation;
@@ -72,11 +73,14 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
 
     private static final int USER_ID_NOT_SPECIFIED = -1;
     protected static final int NOTIF_BASE_ID = 4273;
+    protected static final int PROGRESS_NOTIF_ID = 4274;
+    protected static final int ERROR_NOTIF_ID = 4275;
     private static final String TAG = "RecordingService";
     private static final String CHANNEL_ID = "screen_record";
     private static final String GROUP_KEY = "screen_record_saved";
     private static final String EXTRA_RESULT_CODE = "extra_resultCode";
     protected static final String EXTRA_PATH = "extra_path";
+    protected static final String EXTRA_ID = "extra_id";
     private static final String EXTRA_AUDIO_SOURCE = "extra_useAudio";
     private static final String EXTRA_SHOW_TAPS = "extra_showTaps";
     private static final String EXTRA_CAPTURE_TARGET = "extra_captureTarget";
@@ -259,7 +263,9 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                     startActivity(Intent.createChooser(shareIntent, shareLabel)
                             .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
                     // Remove notification
-                    mNotificationManager.cancelAsUser(null, mNotificationId, currentUser);
+                    final int id = intent.getIntExtra(EXTRA_ID, mNotificationId);
+                    mNotificationManager.cancelAsUser(null, id, currentUser);
+                    maybeDismissGroup(currentUser);
                     return false;
                 }, false, false);
 
@@ -267,9 +273,6 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 closeSystemDialogs();
                 break;
             case ACTION_DELETE:
-                // Close quick shade
-                closeSystemDialogs();
-
                 ContentResolver resolver = getContentResolver();
                 Uri uri = intent.getParcelableExtra(EXTRA_PATH, Uri.class);
                 resolver.delete(uri, null, null);
@@ -280,8 +283,13 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                         Toast.LENGTH_LONG).show();
 
                 // Remove notification
-                mNotificationManager.cancelAsUser(null, mNotificationId, currentUser);
+                final int id = intent.getIntExtra(EXTRA_ID, mNotificationId);
+                mNotificationManager.cancelAsUser(null, id, currentUser);
+                maybeDismissGroup(currentUser);
                 Log.d(TAG, "Deleted recording " + uri);
+
+                // Close quick shade
+                maybeCloseSystemDialogs();
                 break;
         }
         return Service.START_STICKY;
@@ -344,7 +352,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 .setSmallIcon(R.drawable.ic_screenrecord)
                 .setContentTitle(notificationTitle)
                 .addExtras(extras);
-        startForeground(mNotificationId, builder.build());
+        startForeground(ERROR_NOTIF_ID, builder.build());
     }
 
     @VisibleForTesting
@@ -380,7 +388,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 .setForegroundServiceBehavior(Notification.FOREGROUND_SERVICE_IMMEDIATE)
                 .addAction(stopAction)
                 .addExtras(extras);
-        startForeground(mNotificationId, builder.build());
+        startForeground(PROGRESS_NOTIF_ID, builder.build());
     }
 
     @VisibleForTesting
@@ -415,7 +423,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 strings().getShareLabel(),
                 PendingIntent.getService(
                         this,
-                        REQUEST_CODE,
+                        mNotificationId, /* unique request code */
                         getShareIntent(this, uri),
                         PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE))
                 .build();
@@ -425,7 +433,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 getResources().getString(R.string.screenrecord_delete_label),
                 PendingIntent.getService(
                         this,
-                        REQUEST_CODE,
+                        mNotificationId, /* unique request code */
                         getDeleteIntent(this, uri),
                         PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE))
                 .build();
@@ -439,7 +447,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 .setContentText(strings().getSaveText())
                 .setContentIntent(PendingIntent.getActivity(
                         this,
-                        REQUEST_CODE,
+                        mNotificationId, /* unique request code */
                         viewIntent,
                         PendingIntent.FLAG_IMMUTABLE))
                 .addAction(shareAction)
@@ -464,6 +472,9 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
      * grouped together, and the foreground service recording notification is not
      */
     private void postGroupNotification(UserHandle currentUser) {
+        if (countGroupNotifications() < 1) {
+            return; // only post after we show the 2nd notification
+        }
         Bundle extras = new Bundle();
         extras.putString(Notification.EXTRA_SUBSTITUTE_APP_NAME,
                 strings().getTitle());
@@ -472,11 +483,38 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 .setContentTitle(strings().getSaveTitle())
                 .setGroup(GROUP_KEY)
                 .setGroupSummary(true)
+                .setAutoCancel(true)
                 .setExtras(extras)
                 .build();
         mNotificationManager.notifyAsUser(getTag(), NOTIF_BASE_ID, groupNotif, currentUser);
     }
 
+    private void maybeDismissGroup(UserHandle currentUser) {
+        if (countGroupNotifications() >= 1) {
+            return; // dismiss only when we have one notification left
+        }
+        mNotificationManager.cancelAsUser(TAG, NOTIF_BASE_ID, currentUser);
+    }
+
+    private void maybeCloseSystemDialogs() {
+        if (countGroupNotifications() > 0) {
+            return; // only dismiss when we cancel the last group notification
+        }
+        closeSystemDialogs();
+    }
+
+    private int countGroupNotifications() {
+        StatusBarNotification[] notifications = mNotificationManager.getActiveNotifications();
+        int count = 0;
+        for (StatusBarNotification notification : notifications) {
+            final int id = notification.getId();
+            if (id != NOTIF_BASE_ID && id != PROGRESS_NOTIF_ID && id != ERROR_NOTIF_ID) {
+                count++;
+            }
+        }
+        return count;
+    }
+
     private void stopService() {
         stopService(USER_ID_NOT_SPECIFIED);
     }
@@ -519,22 +557,24 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
 
     private void saveRecording(int userId) {
         UserHandle currentUser = new UserHandle(userId);
-        mNotificationManager.notifyAsUser(null, mNotificationId,
+        mNotificationManager.notifyAsUser(null, PROGRESS_NOTIF_ID,
                 createProcessingNotification(), currentUser);
 
         mLongExecutor.execute(() -> {
             try {
                 Log.d(getTag(), "saving recording");
+                postGroupNotification(currentUser);
                 Notification notification = createSaveNotification(
                         getRecorder() != null ? getRecorder().save() : null);
-                postGroupNotification(currentUser);
-                mNotificationManager.notifyAsUser(null, mNotificationId,  notification,
+                mNotificationManager.cancelAsUser(null, PROGRESS_NOTIF_ID, currentUser);
+                mNotificationManager.notifyAsUser(null, mNotificationId, notification,
                         currentUser);
             } catch (IOException | IllegalStateException e) {
                 Log.e(getTag(), "Error saving screen recording: " + e.getMessage());
                 e.printStackTrace();
                 showErrorToast(R.string.screenrecord_save_error);
-                mNotificationManager.cancelAsUser(null, mNotificationId, currentUser);
+                mNotificationManager.cancelAsUser(null, PROGRESS_NOTIF_ID, currentUser);
+                maybeDismissGroup(currentUser);
             }
         });
     }
@@ -683,13 +723,23 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
     }
 
     private Intent getShareIntent(Context context, Uri path) {
-        return new Intent(context, this.getClass()).setAction(ACTION_SHARE)
-                .putExtra(EXTRA_PATH, path);
+        return getShareIntent(context, path, mNotificationId);
+    }
+
+    private Intent getShareIntent(Context context, Uri path, int id) {
+        return new Intent(context, RecordingService.class).setAction(ACTION_SHARE)
+                .putExtra(EXTRA_PATH, path)
+                .putExtra(EXTRA_ID, id);
     }
 
     private Intent getDeleteIntent(Context context, Uri path) {
-        return new Intent(context, this.getClass()).setAction(ACTION_DELETE)
-                .putExtra(EXTRA_PATH, path);
+        return getDeleteIntent(context, path, mNotificationId);
+    }
+
+    private Intent getDeleteIntent(Context context, Uri path, int id) {
+        return new Intent(context, RecordingService.class).setAction(ACTION_DELETE)
+                .putExtra(EXTRA_PATH, path)
+                .putExtra(EXTRA_ID, id);
     }
 
     @Override
-- 
2.47.1

